### PR TITLE
Enable Conv fusion optimizations in optimizeForIdeep

### PR DESCRIPTION
**Pull request by gujinghui**
Enable fusion for IDEEP in optimizeForIdeep
including Conv+ReLU, Conv+Sum, Conv+Sum+ReLU, Conv+BN

### CODE REVIEW

**Comment by gujinghui** [comment]
@yinghai @jgong5

File: caffe2/opt/converter.cc [hunk 1]
```diff
@@ -101,6 +101,14 @@ std::vector<int> getKernelShape(std::map<std::string, caffe2::Argument> argMap)
   return kernelShape;
 }
 
+int getGroup(std::map<std::string, caffe2::Argument> argMap) {
```
Comment (yinghai) [inline caffe2/opt/converter.cc]: pass by value?

File: caffe2/opt/optimize_ideep.cc [hunk 2]
```diff
@@ -1,51 +1,377 @@
...
+    auto newOutput = nn->dataFlow.createNode(
+        unique_dyn_cast<repr::NeuralNetData>(newOutputTensor));
+    nn->dataFlow.replaceNode(convOutput, newOutput);
+
+    nn->dataFlow.deleteNode(convOutput);
+  }
+}
+
+void OptimizeForIdeep(repr::NNModule* nn, caffe2::Workspace* ws, bool training_mode) {
+  if (training_mode) {
```
Comment (yinghai) [inline caffe2/opt/optimize_ideep.cc]: Having `training_mode` will make it more difficult to enforce a uniform optimization API. I'd suggest going though conv operators and see if any of the `training_mode` attribute is set.

**Review by yinghai — COMMENTED** [review]
Thanks for splitting it. Looks good overall. I have 2 minor comments regarding the interface.

File: caffe2/opt/optimize_ideep.cc (hunk 2 above)
Comment (gujinghui) [reply caffe2/opt/optimize_ideep.cc]: The conv-training_mode makes ideep device incompatibale with general pb file. We'd better donot make optimizeForIdeep depends on it. Or just check conv-training_mode optionally.

File: caffe2/opt/converter.cc (hunk 1 above)
Comment (gujinghui) [reply caffe2/opt/converter.cc]: suppose to give a default value to group field.
BTW, can we give default values to other fields, like dilations, pads... ?

**Comment by gujinghui** [comment]
@yinghai   any other concerns?

File: caffe2/opt/optimize_ideep.cc [hunk 3]
```diff
@@ -1,51 +1,377 @@
...
+  return nullptr;
+}
+
+const caffe2::OperatorDef* getOpDef(const repr::NeuralNetOperator& nnOp) {
+  auto annotation = nnOp.getAnnotation();
+  if (!annotation || !isa<Caffe2Annotation>(annotation)) {
+    return nullptr;
+  }
+
+  return dyn_cast<Caffe2Annotation>(annotation)->getOperatorDef();
```
Comment (yinghai) [inline caffe2/opt/optimize_ideep.cc]: It no longer returns pointer. Please rebase against recent master.

File: caffe2/opt/optimize_ideep.cc [hunk 4]
```diff
@@ -1,51 +1,377 @@
...
     }
 
-    // IDEEP doesn't support fusion group conv
-    int group =
-        ArgumentHelper::GetSingleArgument<OperatorDef, int>(op, "group", 1);
-    if (group != 1) {
-      return false;
+    auto convOutput = repr::nn::getOutputs(convNode).front();
+    auto consumers = repr::nn::getConsumers(convOutput);
+    // convOutput is NOT referenced by sequencial ops after BN.
```
Comment (yinghai) [inline caffe2/opt/optimize_ideep.cc]: sequential.

File: caffe2/opt/optimize_ideep.cc [hunk 5]
```diff
@@ -1,51 +1,377 @@
...
+
+    bool should_fuse = true;
+    for (auto input : sumInputs) {
+      auto consumer = repr::nn::getConsumers(input).back();
+      if (consumer != sumNode) {
+        should_fuse = false;
+        break;
+      }
+    }
+    // Sum inputs should not be referenced by sequencial ops.
```
Comment (yinghai) [inline caffe2/opt/optimize_ideep.cc]: same spell here.

File: caffe2/opt/converter.cc [hunk 6]
```diff
@@ -101,6 +101,14 @@ std::vector<int> getKernelShape(std::map<std::string, caffe2::Argument> argMap)
...
+  if (argMap.count("group")) {
+    assert(argMap["group"].has_i() && "Invalid group argument");
```
Comment (yinghai) [inline caffe2/opt/converter.cc]: Use `CAFFE_ENFORCE` instead.

File: caffe2/opt/converter.cc (hunk 1 above)
Comment (yinghai) [reply caffe2/opt/converter.cc]: I don't understand. Why can't we use `const std::map<std::string, caffe2::Argument>&` here?

**Review by yinghai — CHANGES_REQUESTED** [review]
Please clang-format your code.

File: caffe2/opt/converter.cc (hunk 1 above)
Comment (gujinghui) [reply caffe2/opt/converter.cc]: Looks like we have different code base. Will fix it soon.

File: caffe2/opt/converter.cc (hunk 6 above)
Comment (gujinghui) [reply caffe2/opt/converter.cc]: code base changed.... will fix it.

File: caffe2/opt/optimize_ideep.cc (hunk 4 above)
Comment (gujinghui) [reply caffe2/opt/optimize_ideep.cc]: fixed

File: caffe2/opt/optimize_ideep.cc (hunk 5 above)
Comment (gujinghui) [reply caffe2/opt/optimize_ideep.cc]: fixed

File: caffe2/opt/optimize_ideep.cc (hunk 3 above)
Comment (gujinghui) [reply caffe2/opt/optimize_ideep.cc]: fixed

**Comment by gujinghui** [comment]
@yinghai 
code in convert.cc all use 'pass by value'.
Or, we have totally different code base?
Or, we'd better fix all of them...

**Review by yinghai — APPROVED** [review]
Let's have a separate pass to fix the pass-by-value issue.

**Comment by gujinghui** [comment]
@yinghai 
This PR is under evaluation? When will it be merged?
Another patch depends on this one. Pls inform me if this is merged.
Thanks.

File: caffe2/opt/converter.cc [hunk 7]
```diff
@@ -42,6 +42,14 @@ std::vector<int> getDilations(std::map<std::string, caffe2::Argument> argMap) {
   return dilations;
 }
 
+int getGroup(std::map<std::string, caffe2::Argument>& argMap) {
+  if (argMap.count("group")) {
+    CAFFE_ENFORCE(argMap["group"].has_i() && "Invalid group argument");
```
Comment (yinghai) [inline caffe2/opt/converter.cc]: `CAFFE_ENFORCE(argMap["group"].has_i(), "Invalid group argument");`

File: caffe2/opt/optimize_ideep.cc [hunk 8]
```diff
@@ -1,51 +1,381 @@
...
+                      bool training_mode) {
+  LOG(WARNING) << "Only support optimizations for IDEEP";
+}
+
+#else
+USE_IDEEP_DEF_ALIASES();
+
+Blob *getBlob(repr::NNGraph::NodeRef node, caffe2::Workspace *ws) {
+  auto tensor = repr::nn::get<repr::Tensor>(node);
+  assert(ws->HasBlob(tensor->getName()) && "Blob not in workspace");
```
Comment (yinghai) [inline caffe2/opt/optimize_ideep.cc]: Let's use `CAFFE_ENFORCE` everywhere

**Review by yinghai — CHANGES_REQUESTED** [review]
Sorry, Let's use `CAFFE_ENFORCE`

**Comment by gujinghui** [comment]
@yinghai 
fixed

**Review by yinghai — APPROVED** [review]

**Pull request closed without merging**